src/v/storage/key_offset_map.cc

```diff
@@ -49,4 +49,182 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+        const auto [stop, value] = handle_entry(index.value());
+        if (stop) {
+            return seastar::make_ready_future<std::optional<model::offset>>(
+              value);
+        }
+        const auto next_index = probe.next();
+        if (next_index.has_value()) {
+            index = next_index;
+        } else {
+            break;
+        }
+    }
+
+    // fall back to linear probe
+    const auto linear_base = index.value_or(0);
+    for (size_t probe = 0; probe < entries_.size(); ++probe) {
+        const auto [stop, value] = handle_entry(linear_base + probe);
+        if (stop) {
+            return seastar::make_ready_future<std::optional<model::offset>>(
+              value);
+        }
+    }
+
+    return seastar::make_ready_future<std::optional<model::offset>>(
+      std::nullopt);
+}
+
+seastar::future<bool>
+hash_key_offset_map::put(const compaction_key& key, model::offset offset) {
+    const auto hash = hash_key(key);
+
+    // handle a non-normalized probe position
+    // returns true if key is inserted
+    auto handle_entry = [&](size_t index) {
+        ++probe_count_;
+        index = index % entries_.size();
+        auto& entry = entries_[index];
+        if (entry.empty()) {
+            entry.digest = hash;
+            entry.offset = offset;
+            ++size_;
+            max_offset_ = std::max(max_offset_, offset);
+            return true;
+        } else if (entry.digest == hash) {
+            if (offset > entry.offset) {
+                entry.offset = offset;
+                max_offset_ = std::max(max_offset_, offset);
+            }
+            return true;
+        }
+        return false;
+    };
+
+    ++search_count_;
+
+    // probe using the hash
+    probe probe(hash);
+    auto index = probe.next();
+    while (index.has_value()) {
+        const auto inserted = handle_entry(index.value());
+        if (inserted) {
+            return seastar::make_ready_future<bool>(inserted);
+        }
+        const auto next_index = probe.next();
+        if (next_index.has_value()) {
+            index = next_index;
+        } else {
+            break;
+        }
+    }
+
+    // fall back to linear probe
+    const auto linear_base = index.value_or(0);
+    for (size_t probe = 0; probe < entries_.size(); ++probe) {
+        const auto inserted = handle_entry(linear_base + probe);
+        if (inserted) {
+            return seastar::make_ready_future<bool>(inserted);
+        }
+    }
+
+    return seastar::make_ready_future<bool>(false);
+}
+
+model::offset hash_key_offset_map::max_offset() const { return max_offset_; }
+
+size_t hash_key_offset_map::size() const { return size_; }
+size_t hash_key_offset_map::capacity() const { return capacity_; }
+
+seastar::future<> hash_key_offset_map::reset(size_t size) {
+    while (!entries_.empty()) {
+        entries_.pop_back_n(
+          std::min(entries_.size(), entries_.elements_per_fragment()));
+        if (seastar::need_preempt()) {
```

Comment:
this is cool.

Reply:
Yeh. Seems to address the performance cost of the co_await in a tight loop.